Add support for RuntimeConfig API.

tswast opened the pull request:
Supports listing and getting variables, but not the other methods of the RuntimeConfig API, since
I didn't need those, yet.

First step of https://github.com/GoogleCloudPlatform/google-cloud-python/issues/2476

dhermes commented on runtimeconfig/README.rst:
```diff
@@ -0,0 +1,47 @@
+-  `API Documentation`_
+
+.. _Homepage: https://googlecloudplatform.github.io/google-cloud-python/
+.. _API Documentation: http://googlecloudplatform.github.io/google-cloud-python/
```
These links are the same

tswast replied:
Ah, yeah I see that now. They appear to be the same in a bunch of locations:

- https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/datastore/README.rst
- https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/vision/README.rst

I'll delete the "Homepage" link.

dhermes replied:
Good find! Thanks

dhermes commented on runtimeconfig/README.rst:
```diff
@@ -0,0 +1,47 @@
+The Google Cloud `RuntimeConfig`_ (`RuntimeConfig API docs`_) API enables
+developers to dynamically configure and expose variables through Google Cloud
+Platform. In addition, you can also set Watchers and Waiters (not yet supported
+by this client library) that will watch for changes to your data and return
```
Ditch the "(not yet supported by this client library)"

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2014 Google Inc.
```
s/2014/2016

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# limitations under the License.
+
+"""Google Cloud Runtime Configurator API package."""
+from google.cloud.runtimeconfig.client import Client
```
Please add an empty line between the module docstring and the imports

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/_helpers.py:
```diff
@@ -0,0 +1,34 @@
+    """Extract the config name from a full resource name.
+    
+      >>> _config_name_from_full_name('projects/my-proj/configs/my-config')
+      "my-config"
```
Please provide documentation for the argument and the return type

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/_helpers.py:
```diff
@@ -0,0 +1,34 @@
+      >>> _config_name_from_full_name('projects/my-proj/configs/my-config')
+      "my-config"
+    """
+    return full_name.split('/')[3]
```
This is pretty brittle. If you want to assert there are 4 parts you should do the following:

```python
_, _, _, result = full_name.split('/')
return result
```

But it's unclear that there would also be 4 parts. You may be happier just using

```python
separator = '/'
if separator in full_name:
    _, result = full_name.rsplit(separator, 1)
else:
    result = full_name
```

tswast replied:
Okay, I do want to assert that there are 4 parts, so I'll use that.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/_helpers.py:
```diff
@@ -0,0 +1,34 @@
+      >>> _variable_name_from_full_name('projects/my-proj/configs/my-config/variables/another/var/name')
+      "another/var/name"
+    """
+    return '/'.join(full_name.split('/')[5:])
```
I'm also uncomfortable with the `[5:]`. Can you be more explicit about what is being split?

tswast replied:
Added some comments and asserts. It's a bit harder with variable names, since they can use / characters.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+from google.cloud._helpers import _rfc3339_to_datetime
+from google.cloud.exceptions import NotFound
+from google.cloud.iterator import Iterator
+from google.cloud.runtimeconfig._helpers import _config_name_from_full_name, _variable_name_from_full_name
```
Put these imports on separate lines (no commas in imports)

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/_helpers.py:
```diff
@@ -0,0 +1,34 @@
+    """
+    return full_name.split('/')[3]
+
+def _variable_name_from_full_name(full_name):
```
Since the `_helpers` module is already non-public, this function name can be public within the module

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+        self._properties = {}
+
+    def __repr__(self):
+        return '<Config: %s>' % self.name
```
Always pass a tuple to `%`:

```python
 return '<Config: %s>' % (self.name,)
```

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+        ``projects/my-project/configs/my-config``
+
+        :rtype: string
+        :returns: The full name based on project and config names.
```
Add a `raises` section

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+        :rtype: string
+        :returns: The URL path based on project and config names.
+        """
+        return '/%s' % (self.full_name)
```
`(self.full_name)` is identical to `self.full_name`. You want a comma in there to make it a tuple.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+
+"""Create / interact with Google Cloud RuntimeConfig configs."""
+
+import copy
```
Do you use this import?

tswast replied:
Removed. Sorry about these. I should have run flake8 before mailing the PR.

dhermes replied:
No worries. I recommend reading https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/CONTRIBUTING.rst

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+
+import copy
+
+import six
```
Do you use this import?

tswast replied:
Removed.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+            client.connection.api_request(method='GET', path=self.path,
+                                          query_params=query_params,
+                                          _target_object=None)
+            # NOTE: This will not fail immediately in a batch. However, when
```
You don't define a batch in this PR. This note is copy-pasta. Please remove.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+            # minimize the returned payload.
+            query_params = {'fields': 'name'}
+            # We intentionally pass `_target_object=None` since fields=name
+            # would limit the local properties.
```
`_target_object` is only relevant to the storage Batch implementation, please remove this comment and remove `_target_object=None` below

tswast replied:
Ah, okay. Yep, I copied this from the Storage classes. Removed here and elsewhere.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+                method='GET', path=variable.path, _target_object=variable)
+            # NOTE: We assume response.get('name') matches `variable_name`.
+            variable._set_properties(api_response=response)
+            # NOTE: This will not fail immediately in a batch. However, when
```
Remove this Batch reference note (as above) here and below

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+            return None
+
+    def list_variables(self, page_size=None, page_token=None, filter=None,
+                   fields=None, client=None):
```
Bad indent

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+                           passed, will return the first page of variables.
+
+        :type filter: string or ``NoneType``
+        :param filter: optional filter used to filter variables.
```
Please rename to `filter_`. The name `filter` is a Python `__builtin__`

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+            params['pageToken'] = page_token
+
+        if filter is not None:
+            params['filter'] = filter
```
As you can see from the syntax highlighting, `filter` has special meaning in Python

tswast replied:
Done. Yeah, I know it's a built-in, but didn't realize we didn't want to alias them.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+        if fields is not None:
+            params['fields'] = fields
+
+        path = '%s/variables' % (self.path)
```
`(self.path)` --> `(self.path,)`

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,291 @@
+        resp = connection.api_request(method='GET', path=path,
+                                      query_params=params)
+        variables = []
+        for item in resp.get('variables', []):
```
`resp.get('variables', [])` --> `resp.get('variables', ())`

tswast replied:
Done. Does `()` evaluate to an empty tuple? That's kind of surprising, I'd figure I'd need a comma.

dhermes replied:
Yes it's an empty tuple.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,204 @@
+        ``projects/my-project/configs/my-config/variables/my-var``
+
+        :rtype: string
+        :returns: The full name based on config and variable names.
```
Add `raises` section

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,204 @@
+                  ``None`` if the property is not set locally.
+        """
+        if 'updateTime' in self._properties:
+            return _rfc3339_to_datetime(self._properties['updateTime'])
```
This code does two `O(1)` lookups, but could instead just use `self._properties.get('updateTime')` and then check if `None` after the fact

tswast replied:
Done.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,204 @@
+            # minimize the returned payload.
+            query_params = {'fields': 'name'}
+            # We intentionally pass `_target_object=None` since fields=name
+            # would limit the local properties.
```
Remove comment

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,204 @@
+                                          query_params=query_params,
+                                          _target_object=None)
+            # NOTE: This will not fail immediately in a batch. However, when
+            #       Batch.finish() is called, the resulting `NotFound` will be
```
Remove comment

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,204 @@
+            # would limit the local properties.
+            client.connection.api_request(method='GET', path=self.path,
+                                          query_params=query_params,
+                                          _target_object=None)
```
Don't use `_target_object`

dhermes commented on runtimeconfig/tox.ini:
```diff
@@ -0,0 +1,30 @@
+    pytest
+covercmd =
+    py.test --quiet \
+      --cov=google.cloud.vision \
```
Not `vision`

dhermes commented on runtimeconfig/unit_tests/test__helpers.py:
```diff
@@ -0,0 +1,60 @@
+        config_name = self._callFUT(PATH)
+        self.assertEqual(config_name, CONFIG_NAME)
+
+class Test_variable_name_from_full_name(unittest.TestCase):
```
Two empty lines between top-level definitions

dhermes commented on runtimeconfig/unit_tests/test__helpers.py:
```diff
@@ -0,0 +1,60 @@
+        CONFIG_NAME = 'CONFIG_NAME'
+        PROJECT = 'my-project-1234'
+        PATH = 'projects/%s/configs/%s/variables/%s' % (
+                PROJECT, CONFIG_NAME, VARIABLE_NAME)
```
4-space indent here (not 8)

dhermes commented on runtimeconfig/unit_tests/test__helpers.py:
```diff
@@ -0,0 +1,60 @@
+        CONFIG_NAME = 'CONFIG_NAME'
+        PROJECT = 'my-project-1234'
+        PATH = 'projects/%s/configs/%s/variables/%s' % (
+                PROJECT, CONFIG_NAME, VARIABLE_NAME)
```
Ditto

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,204 @@
+
+import httplib2
+import six
+from six.moves.urllib.parse import quote
```
None of these imports is ever used. What gives?

tswast replied:
Removed.

dhermes commented on runtimeconfig/unit_tests/test_config.py:
```diff
@@ -0,0 +1,290 @@
+    def test_ctor(self):
+        client = _Client(project=self.PROJECT)
+        config = self._makeOne(name=self.CONFIG_NAME,
+                              client=client)
```
Bad indent

tswast commented:
Oops, looks like I mixed tabs and spaces and didn't reach 100% coverage on variable.py. Addressing now.

tswast commented:
Updated, please take a second look.

tswast commented:
I'm getting a lint error

```
************* Module google.cloud.runtimeconfig._helpers
C: 52, 0: Line too long (89/80) (line-too-long)
```

The line is

```
                      'projects/project-name/configs/config-name/variables/variable-name'
```

I don't think there's anything I can do to shrink it down.

tswast commented:
Okay, I think I finally made the linter happy. I shrunk the problematic line down by changing the "names" project-name -> proj-name, config-name -> cfg-name

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/_helpers.py:
```diff
@@ -0,0 +1,67 @@
+
+    # Variables are hierarchical. Each node in the hierarchy is separated by a
+    # / character.
+    return '/'.join(parts[5:])
```
Big :-1: on the asserts. Instead try the following:

```python
projects, _, configs, _, variables, result = full_name.split('/', 5)
if (projects != 'projects' or configs != 'configs' or
        variables != 'variables'):
    raise ValueError(
        'Unexpected format of resource', full_name,
        'Expected "projects/{proj}/configs/{cfg}/variables/..."')
return result
```

tswast replied:
Done. I added similar checks to the config name extract function, too.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,290 @@
+        :rtype: string
+        :returns: The full name based on project and config names.
+
+        :raises: :class:`ValueError` if the config is missing a name.
```
Typically `rtype/returns/raises` can go all together with no empty lines. No need to change it, just future reference.

tswast replied:
Acknowledged.

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,197 @@
+                  ``None`` if the property is not set locally.
+        """
+        value = self._properties.get('updateTime')
+        if value:
```
Use an explicit `None` check instead: `if value is not None`

From [PEP8][1]

> Also, beware of writing `if x` when you really mean `if x is not None` -- e.g. when testing whether a variable or argument that defaults to `None` was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!

[1]: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

dhermes commented:
@tswast RE: a line with a string that's too long (like `'projects/project-name/configs/config-name/variables/variable-name'`)

You can use an implicit strong concat in Python:

```python
my_var = ('the beginning of the string '
          'and here is some more of it')
```

----

RE: Mixing tabs and spaces, please get rid of all tabs. You can check for them via `git grep -l $'\t'`. If any files have tabs other than

```
Makefile.bigtable_v2
Makefile.datastore
docs/Makefile
docs/_static/images/favicon.ico
docs/_static/images/gcp-logo-32x32.png
docs/_static/images/gcp-logo.png
system_tests/data/CloudPlatform_128px_Retina.png
system_tests/data/five-point-one-mb-file.zip
system_tests/key.json.enc
```

then you should find the offending lines via `git grep -n $'\t'` and get rid of the tabs

dhermes commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,197 @@
+        except NotFound:
+            return False
+
+    def reload(self, client=None, fields=None):
```
Why did you add `fields=None` here?

tswast replied:
Because I think it will be useful to to a GET request and be able to provide which fields to load. When I was increasing the coverage, I found it strange that I could provide the fields argument to list_variables but not get_variable.

dhermes replied:
1. It will be useful
1. Adding a feature mid-review isn't a great idea. You can upgrade features outside of this PR. You don't need to do the entire implementation in one PR. That makes it harder on reviewers and greatly increases the likelihood of "something missed".
1. You are just provided `fields` the way the backend asks for it, but there is a more user friendly way: take an iterable and then create the `fields` string for the user.

tswast replied:
Since I was copying the storage sample, I got the fields behavior from here:
https://github.com/GoogleCloudPlatform/google-cloud-python/blob/a7cb215d1fb54e00299778d5d04f55893a5f0c9f/storage/google/cloud/storage/bucket.py#L271

and here:
https://github.com/GoogleCloudPlatform/google-cloud-python/blob/5e3a35e63474e56df39f906e71a2746f390f5b5e/storage/google/cloud/storage/client.py#L246

Do we have an example of not passing it through directly?

(The reason I added this parameter is I didn't know how to cover the branches where the variable properties are not set in a way that makes sense. Without passing the fields value to the API, and updateTime will always be present, for example)

tswast replied:
Makes sense. I've removed the `fields` parameters in my latest commit. I also removed `filter_` from `list_variables` since I don't actually know what that parameter does. I'd prefer to leave that out until it's needed, too.

dhermes replied:
RE: `filter_`. That is fine with me. I'm deferring to you as the expert on this particular API.

dhermes replied:
Thanks for punting on `fields`.

dhermes commented on runtimeconfig/unit_tests/test_config.py:
```diff
@@ -0,0 +1,326 @@
+    def test_ctor_w_no_name(self):
+        client = _Client(project=self.PROJECT)
+        config = self._makeOne(name=None, client=client)
+        self.assertRaises(ValueError, lambda: config.full_name)
```
Instead, use a context manager:

```python
with self.assertRaises(ValueError):
    config.full_name
```

tswast replied:
Done. I had to do `_ = config.full_name` to make the linter happy.

dhermes replied:
`getattr(config, 'full_name')` would've been preferred.

Do you recall the lint error?

tswast replied:
The lint error was

```
************* Module unit_tests.test_config
W: 54,12: Statement seems to have no effect (pointless-statement)
************* Module unit_tests.test_variable
W: 70,12: Statement seems to have no effect (pointless-statement)
```

dhermes replied:
Lulz. Calling it `pointless` is quite bold of Pylint.

dhermes commented on runtimeconfig/unit_tests/test_variable.py:
```diff
@@ -0,0 +1,188 @@
+        client = _Client(project=self.PROJECT)
+        config = Config(name=self.CONFIG_NAME, client=client)
+        variable = self._makeOne(name=None, config=config)
+        self.assertRaises(ValueError, lambda: variable.full_name)
```
Ditto here on context manager preferred over `lambda`

tswast replied:
Done.

tswast commented:
Ping. Please take a second look. I believe I have addressed all open comments.

tswast commented:
@dhermes @jonparrott Please review. Is this ready to merge? I just rebased against the latest master.

jonparrott commented on runtimeconfig/README.rst:
```diff
@@ -0,0 +1,44 @@
+See the ``google-cloud-python`` API `RuntimeConfig documentation`_ to learn
+how to analyze images using this library.
+
+.. _Vision documentation: https://google-cloud-python.readthedocs.io/en/stable/runtimeconfig-usage.html
```
s/Vision/RuntimeConfig

tseaver replied:
The prose above needs to be adjusted, as well (copy-pasta from `vision`).

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/client.py:
```diff
@@ -0,0 +1,59 @@
+          This will not make an HTTP request; it simply instantiates
+          a config object owned by this client.
+
+        :type config_name: string
```
`str` is a type, `string` is a module with string helpers.

(I have to remind myself of this constantly)

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        self._properties = {}
+
+    def __repr__(self):
+        return '<Config: %s>' % (self.name,)
```
prefer `.format`, in this case:

```
return '<Config: {name!r}>'.format(name=self.name)
```

tseaver replied:
@jonparrott We have been using `%` throughout the `google-cloud-python` codebase:  I'd say leave as-is, and deal with a conversion to `format()` as a separate (and debatable) issue.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        :rtype: string
+        :returns: The URL path based on project and config names.
+        """
+        return '/%s' % (self.full_name,)
```
Prefer format, in this case `'/{full_name}'.format(full_name=self.full_name)` or `'/{}'.format(full_name)` if you want.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+          This will not make an HTTP request; it simply instantiates
+          a variable object owned by this config.
+
+        :type variable_name: string
```
`str`.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+    def path(self):
+        """URL path for the config's APIs.
+
+        :rtype: string
```
`str`.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        Example:
+        ``projects/my-project/configs/my-config``
+
+        :rtype: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        See:
+        https://cloud.google.com/deployment-manager/runtime-configurator/reference/rest/v1beta1/projects.configs#resource-runtimeconfig
+
+        :rtype: string, or ``NoneType``
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+    def project(self):
+        """Project bound to the config.
+
+        :rtype: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+    :param client: A client which holds credentials and project configuration
+                   for the config (which requires a project).
+
+    :type name: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+            # We only need the status code (200 or not) so we seek to
+            # minimize the returned payload.
+            query_params = {'fields': 'name'}
+            client.connection.api_request(method='GET', path=self.path,
```
nit: I prefer to break after the opening `(` to avoid hanging indents.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        :param client: Optional. The client to use.  If not passed, falls back
+                       to the ``client`` stored on the current config.
+
+        :rtype: boolean
```
`bool`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+          >>> print(config.get_variable('does-not-exist'))
+          None
+
+        :type variable_name: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        See:
+        https://cloud.google.com/deployment-manager/runtime-configurator/reference/rest/v1beta1/projects.configs.variables/list
+
+        :type page_size: integer or ``NoneType``
```
`int`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        :type page_size: integer or ``NoneType``
+        :param page_size: maximum number of variables to return.
+
+        :type page_token: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+        if page_token is not None:
+            params['pageToken'] = page_token
+
+        path = '%s/variables' % (self.path,)
```
prefer format

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/config.py:
```diff
@@ -0,0 +1,260 @@
+            variable = self.variable(name)
+            variable._set_properties(api_response=item)
+            variables.append(variable)
+        return variables, resp.get('nextPageToken')
```
nit: prefer blank newline between block statements.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/connection.py:
```diff
@@ -0,0 +1,47 @@
+    :type http: :class:`httplib2.Http` or class that defines ``request()``.
+    :param http: (Optional) HTTP object to make requests.
+
+    :type api_base_url: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/connection.py:
```diff
@@ -0,0 +1,47 @@
+    API_URL_TEMPLATE = '{api_base_url}/{api_version}{path}'
+    """A template for the URL of a particular API call."""
+
+    SCOPE = ('https://www.googleapis.com/auth/cloud-platform',)
```
This is an overly broad scope, is there a smaller set that will work?

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,188 @@
+    See:
+    https://cloud.google.com/deployment-manager/runtime-configurator/reference/rest/v1beta1/projects.configs.variables
+
+    :type name: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,188 @@
+        Example:
+        ``projects/my-project/configs/my-config/variables/my-var``
+
+        :rtype: string
```
`str`

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,188 @@
+        """
+        if not self.name:
+            raise ValueError('Missing variable name.')
+        return '%s/variables/%s' % (self.config.full_name, self.name)
```
format, please.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,188 @@
+        :rtype: string
+        :returns: The URL path based on config and variable names.
+        """
+        return '/%s' % (self.full_name,)
```
format here as well.

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,188 @@
+    def path(self):
+        """URL path for the variable's APIs.
+
+        :rtype: string
```
`str` (you get the idea at this point I think)

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/variable.py:
```diff
@@ -0,0 +1,188 @@
+        https://cloud.google.com/deployment-manager/runtime-configurator/reference/rest/v1beta1/projects.configs.variables#VariableState
+
+        :rtype: string or ``NoneType``
+        :returns: If set, one of "UPDATED", "DELETED", or
```
Could you make these constants?

jonparrott commented on runtimeconfig/google/cloud/runtimeconfig/_helpers.py:
```diff
@@ -0,0 +1,72 @@
+      "my-config"
+
+    :type full_name: string
+    :param full_name: the full resource name of a config. The full resource
```
This hanging indent is out of control, prefer:

```
:param something:
    blah blah blah blah
    blah blah blah blah
```

in this case.

tswast commented:
@jonparrott @dhermes Please take another look. I believe I have addressed all your comments.

> @tswast there doesn't seem to be any way to create configs or variables, is this intentional because this is just the first step?

Yes. I wanted to keep the initial PR as small as possible to start. Plus, dpebot and the other samples I want to create will just need read access, not write access to RuntimeConfig. After this goes in, we can add creates in, if only so that we can create system tests for this. Right now I'm running the following snippets manually to verify that it does what I want:

```
$ gcloud config configurations activate kittenbot
export GOOGLE_CLOUD_PROJECT=my-project-name
$ python

from google.cloud import runtimeconfig
client = runtimeconfig.Client()
config = client.config('kittenbot')
iterator = config.list_variables()
iterator.update_page()
variables = list(iterator.page)
print(variables)
token = iterator.next_page_token
print(token)

v = config.get_variable('slack-token')
print(v)
```
